htlcswitch/switch.go

```diff
@@ -1031,31 +1031,57 @@ func (s *Switch) extractResult(deobfuscator ErrorDecrypter, n *networkResult,
 	case *lnwire.UpdateFailHTLC:
 		// TODO(yy): construct deobfuscator here to avoid creating it
 		// in paymentLifecycle even for settled HTLCs.
-		paymentErr := s.parseFailedPayment(
+		paymentErr, encryptedErr := s.parseFailedPayment(
 			deobfuscator, attemptID, paymentHash, n.unencrypted,
 			n.isResolution, htlc,
 		)
 
-		return &PaymentResult{
-			Error: paymentErr,
-		}, nil
+		// ElimOption allows us to safely handle both cases of the
+		// optional error.
+		result := fn.ElimOption(
+			paymentErr,
+			// If no decrypted error is present, we use the raw
+			// encrypted error.
+			func() *PaymentResult {
+				return &PaymentResult{
+					EncryptedError: encryptedErr,
+				}
+			},
+			// If a decrypted error is present, we use that.
+			func(pErr error) *PaymentResult {
+				return &PaymentResult{
+					Error: pErr,
+				}
+			},
+		)
 
+		return result, nil
 	default:
 		return nil, fmt.Errorf("received unknown response type: %T",
 			htlc)
 	}
 }
 
-// parseFailedPayment determines the appropriate failure message to return to
-// a user initiated payment. The three cases handled are:
-//  1. An unencrypted failure, which should already plaintext.
-//  2. A resolution from the chain arbitrator, which possibly has no failure
-//     reason attached.
-//  3. A failure from the remote party, which will need to be decrypted using
-//     the payment deobfuscator.
+// parseFailedPayment determines the appropriate failure message for a failed
+// user-initiated payment. It returns a mutually exclusive pair: either a raw,
+// encrypted error blob, or a decrypted error.
+//
+// The cases handled are:
+//  1. An unencrypted failure: This occurs when the payment fails locally before
+//     clearing the first link. The failure reason is already in plaintext.
+//  2. A resolution from the chain arbitrator: This occurs for on-chain
+//     timeouts, which may not have a specific failure reason.
+//  3. A failure from a remote hop: This is the standard case for a multi-hop
+//     payment failure.
+//     - If a deobfuscator is provided, this function will attempt to decrypt
+//     the failure reason.
+//     - If a deobfuscator is NOT provided, the function returns the raw,
+//     onion-encrypted failure blob. This blob is only fully decryptable by the
+//     entity which built the onion packet.
 func (s *Switch) parseFailedPayment(deobfuscator ErrorDecrypter,
 	attemptID uint64, paymentHash lntypes.Hash, unencrypted,
-	isResolution bool, htlc *lnwire.UpdateFailHTLC) error {
+	isResolution bool, htlc *lnwire.UpdateFailHTLC) (fn.Option[error],
+	[]byte) {
```

Comment:
nit: this could be cleaner if we'd use `fn.Either`. For the standard decrypted error case we could use `caseErr := fn.NewLeft[error, []byte]`, `return caseErr(linkError)` for example and then
```golang
		result := fn.ElimEither(
			eitherErr,
			func(pErr error) *PaymentResult {
				return &PaymentResult{
					Error: pErr,
				}
			},
			func(encryptedErr []byte) *PaymentResult {
				return &PaymentResult{
					EncryptedError: encryptedErr,
				}
			},
		)
```

I think that would be better in line with `It returns a mutually exclusive pair: either a raw,
// encrypted error blob, or a decrypted error.`

Reply:
Updated to use `fn.Either` and `fn.ElimEither`. I'm not super confident with the `fn` package so let me know if it looks as you had imagined.